packages/e2e-tests/test/e2e-fle.spec.ts

```diff
@@ -914,6 +913,311 @@ describe('FLE tests', function () {
+              ],
+            });
+            await shell.waitForPrompt();
+
+            // Shared setup for all substring search tests - create collection once
+            await shell.executeLine(`{
+        mode = ${JSON.stringify(mode)};
+        opts = {
+          keyVaultNamespace: '${dbname}.__keyVault',
+          kmsProviders: { local: { key: 'A'.repeat(128) } },
+        };
+
+        // Setup explicit encryption client
+        explicitMongo = Mongo(db.getMongo(), { ...opts, bypassQueryAnalysis: true });
+        if (mode === 'automatic') {
+          autoMongo = Mongo(db.getMongo(), { ...opts });
+          coll = autoMongo.getDB('${dbname}').${testCollection};
+        }
+
+        keyId1 = explicitMongo.getKeyVault().createKey('local');
+        keyId2 = explicitMongo.getKeyVault().createKey('local');
+
+        substringOptions = {
+          strMinQueryLength: 2,
+          strMaxQueryLength: 10,
+        };
+        encryptedFieldOptions = {
+          ...substringOptions,
+          caseSensitive: false,
+          diacriticSensitive: false,
+          contention: 4
+        };
+
+        explicitMongo.getClientEncryption().createEncryptedCollection('${dbname}', '${testCollection}', {
+          provider: 'local',
+          createCollectionOptions: {
+            encryptedFields: {
+              fields: [{
+                keyId: keyId1,
+                path: 'substringData',
+                bsonType: 'string',
+                queries: [{
+                  queryType: 'substringPreview',
+                  strMaxLength: 60,
+                  ...encryptedFieldOptions
+                }]
+              }, {
+                keyId: keyId2,
+                path: 'prefixSuffixData',
+                bsonType: 'string',
+                queries: [{
+                  queryType: 'prefixPreview',
+                  ...encryptedFieldOptions
+                }, {
+                  queryType: 'suffixPreview',
+                  ...encryptedFieldOptions
+                }]
+              }]
+            }
+          }
+        });
+
+        ce = explicitMongo.getClientEncryption();
+        ecoll = explicitMongo.getDB('${dbname}').${testCollection};
+
+        explicitOpts = (details) => ({
+          algorithm: 'TextPreview',
+          contentionFactor: 4,
+          textOptions: { caseSensitive: false, diacriticSensitive: false, ...details }
+        });
+        substringOpts = explicitOpts({ substring: { ...substringOptions, strMaxLength: 60 } });
+        prefixSuffixOpts = explicitOpts({ prefix: substringOptions, suffix: substringOptions });
+        methods = {
+          substring: { queryType: 'substringPreview', keyId: keyId1, options: substringOpts },
+          prefix: { queryType: 'prefixPreview', keyId: keyId2, options: prefixSuffixOpts },
+          suffix: { queryType: 'suffixPreview', keyId: keyId2, options: prefixSuffixOpts }
+        }
+        explicitEncrypt = (method, data) => {
```

Comment:
nice idea with defining the methods, opts variables. a function for explicit, make it much cleaner to follow. I think untangling the different formats was the biggest problem for me.